### DISCUSSION TITLE
Connectors?

### DISCUSSION THREAD

**Discussion by PWPInnovator898** [kriswilk/buzzah#2] [Q&A]
You suggested. “ Solder pins to each driver and add a matching connector to each motor (as pictured above)”
Can you kindly describe or point to the connector you referenced?  In other words, is it that rectangular section on the side of the driver?

**Comment by kriswilk** [comment]
This is one of the details that I haven't documented in detail, mainly because I wish there were a "better way". But here are some closer pics of the best solution I have so far... The first shows the assembled connection between the driver and the actuator. the second and third show the pins I soldered onto the driver and the connector I crimped onto the actuator wires, respectively. The pins are just [breakaway right-angle header](https://www.adafruit.com/product/1540) pins and the connector is a 3-pin female header connector.

Unless you have the right crimp tool, buying [a cable like this](https://www.adafruit.com/product/4935), cutting it in half, and soldering an actuator to each set of wires (only the outer two) might be the way to go.

![20230316_104425](https://user-images.githubusercontent.com/382436/233860129-3065c1f0-848a-486f-8a37-51ede75aec8e.jpg)
![20230423_151014](https://user-images.githubusercontent.com/382436/233860292-e8e178df-0d6e-4214-956f-7bcfbd4688d3.jpg)
![20230423_151041](https://user-images.githubusercontent.com/382436/233860297-04cded53-30f8-415f-8a40-7c4c5d464818.jpg)

**Reply by PWPInnovator898** [reply]
Clever!!  Thanks for the details too. Of all the designs I’ve seen and built, yours is the most straightforward and practical IMHO.  I’ve built a couple glove protos that should plug right in to your electronics too.  I also was fortunate to test and wear the gloves that were part of the Stanford tests a few weeks ago.  What struck me was the intensity of the vibrations and the z-axis movement sensation of the actuator onto the fingertip. I’ll soon be able to compare that vibration to the vibration sensation that comes from your design, once I receive all of the parts, and I finish assembling later this week.  Thanks again for your excellent work here and also for sharing it with the community.

**Reply by kriswilk** [reply]
No problem. Let me know if you have any issues at all. Since delivering the kit to the fellow who'll be testing it, I made a change to the algorithm. It now sends unique patterns to the L/R hands rather than mirroring them. But I haven't tested that myself.... If you notice any strangeness, I'll fix/adjust the firmware.

I'm very interested to hear your comparison to the vibration from the "real" gloves. Hopefully, it's similar.

**Reply by PWPInnovator898** [reply]
Absolutely. Once everything is assembled I’ll send you information on how it compares with the Stanford gloves.

**Reply by PWPInnovator898** [reply]
I hope to get all parts late tomorrow and complete assembling in a few days…
I think it would be good to have both a mirrored and non-mirrored option for the patterns.

**Reply by kriswilk** [reply]
That's an easy option to add back into the code. Once you've assembled and tested the existing algorithm (to make sure it works as-is), just let me know and I'll add a selectable option that can mirror the pattern from outputs 0-3 to 4-7.

**Comment by PWPInnovator898** [comment]
Hello Kris;
All components have been received, and all but the motors have been
assembled.
The RP uC is updated w/ your latest code.
I plan to finish the motor+glove integration next week.
Does the red LED indicate the beginning/end of each 3ON:2OFF:3ON sequence,
per Dr. Tass?
Thanks again!

On Thu, Apr 27, 2023 at 3:07 PM Kris Wilk ***@***.***> wrote:

> That's an easy option to add back into the code. Once you've assembled and
> tested the existing algorithm (to make sure it works as-is), just let me
> know and I'll add a selectable option that can mirror the pattern from
> outputs 0-3 to 4-7.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kriswilk/buzzah/discussions/2#discussioncomment-5747797>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A6TYIZD75DQKQXZPB5VSTHLXDLNZTANCNFSM6AAAAAAXIWSF6A>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks and best regards,
Phillip Callahan

**Reply by kriswilk** [reply]
If you see a red light on the uC board, it suggests there was a code/runtime error. To debug that, I recommend downloading my ORIGINAL code (which I know works) using this link:

https://github.com/kriswilk/buzzah/raw/3feb9b079eefa31e20b6b3e625e70c34185486be/code.py

You'll probably have to right click and "Save as..." to save it to a file as "code.py".

If you copy that version of `code.py` over to the uC board and still get a red light, you may not have the other required library files installed in the right spot. When correctly setup, the file structure on the uC board should be:

- code.py
- lib\adafruit_itertools\__init__.mpy
- lib\adafruit_itertools\adafruit_itertools_extras.mpy
- lib\adafruit_drv2605.mpy
- lib\adafruit_tca9548a.mpy

If that works you should get a green light. Then it means I need to revisit my recent code changes and find out what I broke. :)

If, on the other hand, you still get a red light, something else is amiss..let me know and I can help you to figure it out. Note also that the system should start properly even without the motors connected to the drivers. But you do need all the other connections (mux, 8 driver boards) otherwise the code will fail when trying to communicate with anything that's missing.

**Reply by PWPInnovator898** [reply]
My fault:  I left 4 drivers off on one side of the mux, simply to do a quick test, so that's the issue. 
Thanks for the quick response!

**Reply by kriswilk** [reply]
Ah, OK. Nevertheless, you may want to try the old code file as a comparison to the current code. The only perceptible difference should be that the old one had the pattern mirrored L/R. I'll add that back as a configurable option once I know that the new code is actually working as expected.

**Reply by GrayGhost-su** [reply]
Kris,
    is this link still where we can get your original code that has the random sequences that are mirrored so both left and right hands are given the same finger sequence ? 
https://github.com/kriswilk/buzzah/raw/3feb9b079eefa31e20b6b3e625e70c34185486be/code.py

Trying to finish mine up and would like to try left/right random and left/right random but mirrored.

Thanks
GG

**Reply by kriswilk** [reply]
Yes, that's the original code. However, I've just updated the "dev" branch to include an option for mirroring L/R patterns. The latest dev code.py is always at:

https://raw.githubusercontent.com/kriswilk/buzzah/dev/code.py

I only changed a few lines but must warn you that I didn't test it yet (I no longer have a full set of hardware here) but it ought to work if you simply change the "MIRROR" option in the top section of the code file to either "False" or "True". It defaults to False to preserve existing behaviour. Changing it to "True" should mirror the pattern L/R. This also assumes that port 0 & 4 are the index fingers, 1 & 5 the middle fingers, and so on.

Please try loading that version of code.py and toggle the option either way. Let me know if it's working as expected and if so, I'll push it to the main branch for everyone to use if they so desire.

**Reply by PWPInnovator898** [reply]
Hi Kris;
Thanks for the updates.

RE: I2C mux port-to-finger mapping, is this what your code assumes?
Left index = I2C port 0; Right index = I2C port 4
Left middle = I2C port 1; Right middle = I2C port 5 
Left ring = I2C port 2;  Right ring = I2C port 6
Left pinky = I2C port 3;  Right pinky = I2C port 7;

I used a slightly different map on my h/w, so I'd rather match yours for avoiding future confusion on my part. :-)
I'll also test the new mirror option and let you know,

Thanks for confirming.

**Reply by kriswilk** [reply]
Yes, that's what the new (and old) code assumes for the purposes of mirroring. Well, really it's just mirroring the pairs 0/4, 1/5, 2/6, 3/7. As long as those port pairs lead to the matching fingers on each hand, it should be fine, since the rest is randomized. If you didn't care about mirroring, it would make no difference whatsoever which ports you used as long as 0-3 are one hand and 4-7 are the other (so that no fingers are repeated in any one sequence of 4 vibrations).

**Reply by GrayGhost-su** [reply]
Thanks Kris I will load the dev branch code and give it a go. also Thank you for the port to finger mapping information.

GG

**Reply by PWPInnovator898** [reply]
GG & KW: 
- the updated [code ](https://raw.githubusercontent.com/kriswilk/buzzah/dev/code.py) was tested:  
- works great in both mirrored/non-mirrored modes
- I added a session timer function to shut off vibrations after 60 min (default)
- Let me know where you prefer I post the code additions...

**Reply by kriswilk** [reply]
The conventional way to share a code addition on github would be for you to first "fork" the repo and make modifications to your own copy. Then you can submit a "pull request" which I can review with you to incorporate into this project.

However, in this case you could even just share by cut/paste since the code is so short.

I presume that you use the time.time() function to measure elapsed time and terminate the loop once it exceeds a specified value? In any case, it's definitely a useful addition.

**Reply by PWPInnovator898** [reply]
Thanks Kris;
here is  the code snippet for the session timer:

.....
# Stop VCR therapy session after this number of minutes
THERAPY_TIME = 60

# stop vibration session after THERAPY_TIME in minutes
    t_end = time.time() + 60 * THERAPY_TIME

while time.time() < t_end:
    buzz_sequence(random_sequence())
    buzz_sequence(random_sequence())
    buzz_sequence(random_sequence())
    time.sleep(TIME_RELAX)
    time.sleep(TIME_RELAX)

Thanks again for your helpful work on this!

**Reply by kriswilk** [reply]
I've added this code to the dev branch (with minor adjustment for naming consistency). Please confirm that it's working and I'll push it to main.

Thanks for your contribution!

**Reply by kriswilk** [reply]
Forgot to add that I changed the default session length to 120 minutes to match the published research. I believe that's the latest info but welcome any correction on that.

**Reply by PWPInnovator898** [reply]
The latest updated code works as expected. 
You are correct - 120min is the currently documented session time.
Thanks again for the update!

**Comment by PWPInnovator898** [comment]
Kris;

Does your code make use of the autoresonance feature of the TI  haptic IC?
In other words, is the operating mode of the motor fixed programmatically
at 250 Hz (per Tass) or is it determined by the TI chip setting the
operating frequency at the IC-determined auto-resonant frequency of the
motor?

The reason I'm asking is because the first 245Hz Vybronics motor
(VG0832022D) I've tested at *250Hz *produces very weak vibrations at
250Hz.   The intensity should fall off quickly per the Vybronics datasheet
but much more than I expected, and nowhere near what I experienced w/ the
Stanford gloves.

There are other wideband motors (VG1040003D) that I have tested @ 250 Hz
with very strong vibration intensity that more closely matches Stanford
intensity, but the autoresonance feature of the TI  haptic IC would likely
need to be disabled.  Thanks for your feedback.

Thanks again, Phillip

On Wed, May 3, 2023 at 10:38 AM Kris Wilk ***@***.***> wrote:

> Ah, OK. Nevertheless, you may want to try the old code file as a
> comparison to the current code. The only perceptible difference should be
> that the old one had the pattern mirrored L/R. I'll add that back as a
> configurable option once I know that the new code is actually working as
> expected.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kriswilk/buzzah/discussions/2#discussioncomment-5793440>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A6TYIZGZQJBDKVV3XYAV5PDXEKCYBANCNFSM6AAAAAAXIWSF6A>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks and best regards,
Phillip Callahan

**Reply by kriswilk** [reply]
My implementation uses the autotuning built into the TI DRV2605. This is the default behaviour of the Adafruit software library that is imported, and is a logical/sane default that will "just work" with any LRA that you connect to the drivers. Indeed, it will even work if you connect DIFFERENT actuators to each driver.

The LRAs that I specified have a resonant frequency of 235 Hz. I may be proven wrong, but I suspect that the exact frequency isn't critical. The goal is to stimulate a certain set of nerves, and they appear to have a [rather broad sensitivity range](https://link.springer.com/article/10.3758/BF03208688) between 200-400 Hz. Whether the conscious perception of the vibration is directly linked to its effectiveness, I don't know. If you know more about this, I'm all ears.

To drive the actuators at precisely 250 Hz would require switching off the autoresonance tuner and running the drivers in "PWM" mode. This cannot be achieved with the same simple assembly. Additional wires would need to be connected from the uC to the driver PWM inputs, and the firmware would need to drive the inputs with a 250Hz waveform. As far as I know, there's no way to simply tell the driver what frequency you want manually...that would be lovely.

**Reply by Neurosmith24** [reply]
Jumping into this conversation pretty late, but thought I'd add that at a webinar awhile back Peter Tass did say that 250 hz frequency is less critical than the pattern, saying, "the 250 Hertz is the frequency of the the so-called intra
burst frequency so the system starts and delivers a 250 Hertz burst and then it stops. the important thing is is less the specific frequency but it is the pattern with which we deliver these bursts at different times at different sides that disrupt the abnormal brain syncrony." You can find that at about the 26:17 mark of in this video: https://youtu.be/uDxJkMrYjI0?si=saGAi8AQHBrRDfkW&t=1556

**Reply by kriswilk** [reply]
@Neurosmith24 That is what I suspected. Good to hear it mentioned by him. Thanks for the link!

**Comment by PWPInnovator898** [comment]
Hi Kris;Thank you for your detailed response. I agree with all of your points,   although I have a suggestion on the DRV2605 operating mode that would operate an LRA based on a user specified frequency rather than the auto resonance  mode determined frequency.  According to the TI datasheet,drv2605lPDF Document · 2.7 MBsection 8.3.5.3, using the device in real time playback mode (which I believe you use?), a user can  additionally and optionally bypass the auto resonant derived frequency by setting a bit, and then programming the period for the desired frequency.  To me, and if this works as advertised, this capability would be very valuable as it would allow using a broader array of LRAs, such as the wide band LRAs that I’ve tested, and which have very strongvibrational performance closely matching the experiences I had with the Stanford gloves.  This would also allow VCR users additional flexibility to address patient specific vibration frequency sensitivities.  LRA Coin Vibration Motor - VG1040003Dvybronics.comPlease let me know your thoughts, and thank you sincerely for your contributions to this community!Thanks and best regards, PhillipOn May 4, 2023, at 22:54, Kris Wilk ***@***.***> wrote:﻿
My implementation uses the autotuning built into the TI DRV2605. This is the default behaviour of the Adafruit software library that is imported, and is a logical/sane default that will "just work" with any LRA that you connect to the drivers. Indeed, it will even work if you connect DIFFERENT actuators to each driver.
The LRAs that I specified have a resonant frequency of 235 Hz. I may be proven wrong, but I suspect that the exact frequency isn't critical. The goal is to stimulate a certain set of nerves, and they appear to have a rather broad sensitivity range between 200-400 Hz. Whether the conscious perception of the vibration is directly linked to its effectiveness, I don't know. If you know more about this, I'm all ears.
To drive the actuators at precisely 250 Hz would require switching off the autoresonance tuner and running the drivers in "PWM" mode. This cannot be achieved with the same simple assembly. Additional wires would need to be connected from the uC to the driver PWM inputs, and the firmware would need to drive the inputs with a 250Hz waveform. As far as I know, there's no way to simply tell the driver what frequency you want manually...that would be lovely.

**Reply by kriswilk** [reply]
Thanks for making me look into this a second time! Specifically, the reference to section 8.3.5.3 of the datasheet. As it turns out, the copy of the datasheet that I've been working from is an old one for the "non-L" variant of the DRV2605 and I hadn't even noticed. It doesn't have that section because the feature is specific to the (newer) "L" variant.

So yes, this looks possible. Fortunately, I still have a driver and LRA here that I can test with. I'll try to work on adding settings to select open-loop mode and the desired driving frequency. Probably sometime next week.

One thing to note is that by switching to open-loop mode you sacrifice the "smart-loop" features like automatic overdrive and braking. So there is the possibility for degraded precision of the onset/termination of each vibration in the sequence.

**Reply by Neurosmith24** [reply]
Hi Kris and Phillip,

Planning to start building gloves soon and wanted to confirm that the code has been updated to allow Phillip's suggestion of disabling the auto-resonance  calibration feature and manually set the driver for the resonant frequency of the LRA? (Is this what 'open loop mode' means?).

Phillip you've mentioned the Vybronics VG1040003D in this discussion but also the VLV1040A in a separate discussion. Which of the two have you found better/most closely matches the Stanford gloves?

Thanks

Jeff

**Reply by PWPInnovator898** [reply]
Hello Jeff;
Yes, the code fully supports operating in open loop mode. 
The VLV1040A is it square 10 mm tactor that provides the best stimulation range intensity as far as proof tested so far and as compared to the gloves from one of the early trials that I had access to last year  
Hope this helps. Thank you.

**Reply by kriswilk** [reply]
@Neurosmith24 Sorry for getting back to this thread so late, I had GitHub notifications turned off and just looked in. As @PWPInnovator898 said, yes, the current code implements open-loop mode. In this mode, the driver makes no attempt to automatically set the frequency, and must be configured by the user.

If you use an actuator other than the one specified in the documentation, you must edit the "actuator paramenters" section of the code file to set the actuator type, maximum drive voltage, and resonant frequency (used for LRA only).

You may set the frequency to any value but will get the highest amplitude vibration when it's tuned to the resonant frequency of the actuator.

**Comment by PWPInnovator898** [comment]
Thanks Kris for your willingness to add this feature!
I really appreciate your willingness to include this in your code.
I think it will be quite useful to support a wider array of LRAs and
expected patient-specific vibration frequency/intensity sensitivities.

Bottom line on the tactors I've tested:

   - Subjectively measured, and even when driven off-resonance, Vybronics' 10mm
   wideband 170 Hz
   <https://www.vybronics.com/coin-vibration-motors/lra/v-g1040003d> tactors
   for example (driven at 250 Hz) appear to provide the highest vibration
   intensity that I have tested so far, and can also match the Stanford gloves
   intensity.
   - I've included a 3d-printed 5mm diameter, "button" on top of the
   tractor to help focus vibration energy. photo attached,

Thanks again for your excellent efforts!
Sincerely, Phillip (a PWP)

On Fri, May 5, 2023 at 8:03 AM Kris Wilk ***@***.***> wrote:

> Thanks for making me look into this a second time! Specifically, the
> reference to section 8.3.5.3 of the datasheet. As it turns out, the copy of
> the datasheet that I've been working from is an old one for the "non-L"
> variant of the DRV2605 and I hadn't even noticed. It doesn't have that
> section because the feature is specific to the (newer) "L" variant.
>
> So yes, this looks possible. Fortunately, I still have a driver and LRA
> here that I can test with. I'll try to work on adding settings to select
> open-loop mode and the desired driving frequency. Probably sometime next
> week.
>
> One thing to note is that by switching to open-loop mode you sacrifice the
> "smart-loop" features like automatic overdrive and braking. So there is the
> possibility for degraded precision of the onset/termination of each
> vibration in the sequence.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kriswilk/buzzah/discussions/2#discussioncomment-5816399>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A6TYIZFL4NYXOSSPKHVBG3TXEUCDNANCNFSM6AAAAAAXIWSF6A>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks and best regards,
Phillip Callahan

**Reply by Neurosmith24** [reply]
@PWPInnovator898 I'm curious about the 3d-printed 5mm button you mention. Did you just glue that onto the Vybronics tactor? Not seeing the photo you mention. Can you re-post. FWIW, I've come across a couple of free 3D CAD software options, FreeCAD and TinkerCAD that I'll be playing with so that I can then edit files for tactor holders, this button, etc. Thanks!

**Reply by PWPInnovator898** [reply]
Yes, I glued the 5mm button on the tactor. Pic attached.  

![image](https://github.com/kriswilk/buzzah/assets/128418916/56adc789-5590-4e4f-9b86-7217f80c1c4f)

tinkercad is very easy to use. 
We used tinkercad to design all tactor holders with the exception of the latest. 
It was designed with Fusion360 as it’s much more powerful and harder to use but not free.   Microsoft’s 3D builder is free Btw too.

**Reply by Neurosmith24** [reply]
Thanks!
